lighthouse-core/gather/gatherers/forms.js

```diff
@@ -0,0 +1,116 @@
+/**
+ * @license Copyright 2020 The Lighthouse Authors. All Rights Reserved.
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License. You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
+ */
+'use strict';
+
+const Gatherer = require('./gatherer.js');
+const pageFunctions = require('../../lib/page-functions.js');
+
+/* eslint-env browser, node */
+
+/**
+ * @return {LH.Artifacts['Forms']}
+ */
+/* istanbul ignore next */
+function collectFormElements() {
+  // @ts-ignore - put into scope via stringification
+  const formChildren = getElementsInDocument('textarea, input, label, select'); // eslint-disable-line no-undef
+  const forms = new Map();
+  /** @type { { inputs: LH.Artifacts['FormInputs'][], labels: LH.Artifacts['FormLabels'][] }  } */
+  const formlessObj = {
+    inputs: [],
+    labels: [],
+  };
+  for (const child of formChildren) {
+    const hasForm = child.closest('form');
```

Comment:
I can't find the previous discussion on why `.form` didn't work. Where is that? I think it was because of labels within a form element but without an associated input element don't get associated with the form via `.form`?

I ran the page code [1] on form.html for both `.form` vs `.closest('form')`.

.closest:

![image](https://user-images.githubusercontent.com/4071474/88838162-c34f1800-d19e-11ea-865b-c89a191c9e2b.png)

.form:

![image](https://user-images.githubusercontent.com/4071474/88838192-cd711680-d19e-11ea-9ee8-6009d62ef977.png)

OK, so they are different, and we think the former is correct, but I wanted to verify that.

```
if (child.form !== child.closest('form')) console.log('diff', {
      child,
      'child.form': child.form,
      closest: child.closest('form')
    });
```

![image](https://user-images.githubusercontent.com/4071474/88840201-bbdd3e00-d1a1-11ea-8994-e95a40e5f9ba.png)

I believe the first label here doesn't have `.form` set because the input element doesn't recognize it as its label. I removed the label's `id` and it worked, but only because a label in another part of the page references this input. I guess duplicate id elements is very bad, at least for understanding what's going on. Can you modify all the ids on this page to be unique? perhaps prefix the formless ones with `formless-` or `formless-1-` or w.e. is needed.

The other 3 labels all have a `.form`, even though they aren't nested within a form element. That's because the reference an input element via `for` that is within that form. And to the browser, that's all that matters. Making the ids unique for this page will fix this issue.

I haven't verified it, but might the results be correct if we did this:
```
let parentFormElement = child.form;
    // comment explaining why this case happens ...
    if (child.tagName === 'LABEL' && !parentFormElement) {
      parentFormElement = child.closest('form');
    }
```

with the added benefit that the edge case is make explicit.

-----

[1] just `console.log(expression)` in the gatherer to get this code, then paste in chrome devtools console
```
(() => {
      function getElementsInDocument(selector) {
  const realMatchesFn = window.__ElementMatches || window.Element.prototype.matches;
  /** @type {Array<HTMLElement>} */
  const results = [];

  /** @param {NodeListOf<HTMLElement>} nodes */
  const _findAllElements = nodes => {
    for (let i = 0, el; el = nodes[i]; ++i) {
      if (!selector || realMatchesFn.call(el, selector)) {
        results.push(el);
      }
      // If the element has a shadow root, dig deeper.
      if (el.shadowRoot) {
        _findAllElements(el.shadowRoot.querySelectorAll('*'));
      }
    }
  };
  _findAllElements(document.querySelectorAll('*'));

  return results;
};
      function getNodePath(node) {
  /** @param {Node} node */
  function getNodeIndex(node) {
    let index = 0;
    let prevNode;
    while (prevNode = node.previousSibling) {
      node = prevNode;
      // skip empty text nodes
      if (node.nodeType === Node.TEXT_NODE && node.nodeValue.trim().length === 0) continue;
      index++;
    }
    return index;
  }

  const path = [];
  while (node && node.parentNode) {
    const index = getNodeIndex(node);
    path.push([index, node.nodeName]);
    node = node.parentNode;
  }
  path.reverse();
  return path.join(',');
};
      function getOuterHTMLSnippet(element, ignoreAttrs = [], snippetCharacterLimit = 500) {
  const ATTRIBUTE_CHAR_LIMIT = 75;
  try {
    // ShadowRoots are sometimes passed in; use their hosts' outerHTML.
    if (element instanceof ShadowRoot) {
      element = element.host;
    }

    const clone = element.cloneNode();
    ignoreAttrs.forEach(attribute =>{
      clone.removeAttribute(attribute);
    });
    let charCount = 0;
    for (const attributeName of clone.getAttributeNames()) {
      if (charCount > snippetCharacterLimit) {
        clone.removeAttribute(attributeName);
      } else {
        let attributeValue = clone.getAttribute(attributeName);
        if (attributeValue.length > ATTRIBUTE_CHAR_LIMIT) {
          attributeValue = attributeValue.slice(0, ATTRIBUTE_CHAR_LIMIT - 1) + '…';
          clone.setAttribute(attributeName, attributeValue);
        }
        charCount += attributeName.length + attributeValue.length;
      }
    }

    const reOpeningTag = /^[\s\S]*?>/;
    const [match] = clone.outerHTML.match(reOpeningTag) || [];
    if (match && charCount > snippetCharacterLimit) {
      return match.slice(0, match.length - 1) + ' …>';
    }
    return match || '';
  } catch (_) {
    // As a last resort, fall back to localName.
    return `<${element.localName}>`;
  }
};
      function getNodeLabel(node) {
  // Inline so that audits that import getNodeLabel don't
  // also need to import truncate
  /**
   * @param {string} str
   * @param {number} maxLength
   * @return {string}
   */
  function truncate(str, maxLength) {
    if (str.length <= maxLength) {
      return str;
    }
    return str.slice(0, maxLength - 1) + '…';
  }

  const tagName = node.tagName.toLowerCase();
  // html and body content is too broad to be useful, since they contain all page content
  if (tagName !== 'html' && tagName !== 'body') {
    const nodeLabel = node.innerText || node.getAttribute('alt') || node.getAttribute('aria-label');
    if (nodeLabel) {
      return truncate(nodeLabel, 80);
    } else {
      // If no useful label was found then try to get one from a child.
      // E.g. if an a tag contains an image but no text we want the image alt/aria-label attribute.
      const nodeToUseForLabel = node.querySelector('[alt], [aria-label]');
      if (nodeToUseForLabel) {
        return getNodeLabel(/** @type {HTMLElement} */ (nodeToUseForLabel));
      }
    }
  }
  return tagName;
};
      return (function collectFormElements() {
  // @ts-ignore - put into scope via stringification
  const formChildren = getElementsInDocument('textarea, input, label, select'); // eslint-disable-line no-undef
  const forms = new Map();
  /** @type { { inputs: Array<LH.Artifacts.FormInput>, labels: Array<LH.Artifacts.FormLabel> }  } */
  const formlessObj = {
    inputs: [],
    labels: [],
  };
  for (const child of formChildren) {
    const parentFormElement = child.form;
    const hasForm = !!parentFormElement;
    if (hasForm && !forms.has(parentFormElement)) {
      const newFormObj = {
        attributes: {
          id: parentFormElement.id,
          name: parentFormElement.name,
          autocomplete: parentFormElement.autocomplete,
          // @ts-ignore - put into scope via stringification
          nodeLabel: getNodeLabel(child), // eslint-disable-line no-undef,
          // @ts-ignore - put into scope via stringification
          snippet: getOuterHTMLSnippet(parentFormElement), // eslint-disable-line no-undef
        },
        inputs: [],
        labels: [],
      };
      forms.set(parentFormElement, newFormObj);
    }
    const formObj = hasForm ? forms.get(parentFormElement) : formlessObj;

    if (child instanceof HTMLInputElement || child instanceof HTMLTextAreaElement) {
      const isButton = child instanceof HTMLInputElement &&
      (child.type === 'submit' || child.type === 'button');
      if (isButton) continue;
      formObj.inputs.push({
        id: child.id,
        name: child.name,
        placeholder: child.placeholder,
        autocomplete: child.autocomplete,
        // @ts-ignore - put into scope via stringification
        nodeLabel: getNodeLabel(child), // eslint-disable-line no-undef,
        // @ts-ignore - put into scope via stringification
        snippet: getOuterHTMLSnippet(child), // eslint-disable-line no-undef
      });
    }
    if (child instanceof HTMLSelectElement) {
      formObj.inputs.push({
        id: child.id,
        name: child.name,
        autocomplete: child.autocomplete,
        // @ts-ignore - put into scope via stringification
        nodeLabel: getNodeLabel(child), // eslint-disable-line no-undef,
        // @ts-ignore - put into scope via stringification
        snippet: getOuterHTMLSnippet(child), // eslint-disable-line no-undef
      });
    }
    if (child instanceof HTMLLabelElement) {
      formObj.labels.push({
        id: child.id,
        for: child.htmlFor,
        // @ts-ignore - put into scope via stringification
        nodeLabel: getNodeLabel(child), // eslint-disable-line no-undef,
        // @ts-ignore - put into scope via stringification
        snippet: getOuterHTMLSnippet(child), // eslint-disable-line no-undef
      });
    }
  }

  if (formlessObj.inputs.length > 0 || formlessObj.labels.length > 0) {
    forms.set('formless', {
      inputs: formlessObj.inputs,
      labels: formlessObj.labels,
    });
  }
  return [...forms.values()];
})();
    })()
```

Reply:
> I can't find the previous discussion on why .form didn't work. Where is that?

We had it in chat :)

> I think it was because of labels within a form element but without an associated input element don't get associated with the form via .form

Right-o

https://melodic-class.glitch.me/form-label.html was my demo of when `.form` works

Reply:
Also, why do we even want to collect labels that aren't well-associated with the input element/form? The browser isn't. Screen readers, probably, maybe won't ?

Reply:
@connorjclark I wanted to collect it for future audits that have evaluate floating labels

Reply:
making the labels unique right now

Reply:
@connorjclark so I should use the form attribute and only collect for an edge case?

Reply:
let parentFormElement = child.form;
    // comment explaining why this case happens ...
    if (child.tagName === 'LABEL' && !parentFormElement) {
      parentFormElement = child.closest('form');
    }

wouldn't this edge case trigger for all formless elements? not only the undetectable ones that dont have a form attribute

Reply:
@connorjclark I think it makes a lot of sense to collect labels that aren't associated with anything in a `form-elements` gatherer. axe checks for unlabeled inputs but AFAIK not the other way around of dangling labels :)

Reply:
> @connorjclark I wanted to collect it for future audits that have evaluate floating labels

> @connorjclark I think it makes a lot of sense to collect labels that aren't associated with anything in a form-elements gatherer. axe checks for unlabeled inputs but AFAIK not the other way around of dangling labels :)

OK.

> @connorjclark so I should use the form attribute and only collect for an edge case?

Yes, that's what I think it clearest. `.form` is what the browser itself thinks about the form, which is easy to lean on. And stuff the unclear stuff in an if statement. If the id uniqueness issue has been fixed, I think the two approaches should have the same results.

> wouldn't this edge case trigger for all formless elements? not only the undetectable ones that dont have a form attribute

The condition is checking for just labels.

Reply:
> OK.
> Yes, that's what I think it clearest.

wait @connorjclark I'm a bit confused aren't these contradictory? If we're using `.form` then the labels won't be put into the form in which they belong if they're not correctly associated with an input. 

Aside: I think the duplicate ID problem is a great example of how the output of using `.closest('form')` could've helped identify the right spot to fix :)

> I think the two approaches should have the same results.

If we're talking about `input.form` vs. `input.closest('form')` then the former will end up with all labels that aren't correctly associated with an input will end up in the formless object instead of their parent form, so the results will be different. Maybe we're talking about different things though?

Reply:
> wait @connorjclark I'm a bit confused aren't these contradictory? If we're using .form then the labels won't be put into the form in which they belong if they're not correctly associated with an input. If we're talking about input.form vs. input.closest('form') then the former will end up with all labels that aren't correctly associated with an input will end up in the formless object instead of their parent form, so the results will be different. Maybe we're talking about different things though?

huh? Did you see this?

![image](https://user-images.githubusercontent.com/4071474/89223512-874ff480-d59c-11ea-80a5-aa59bac364e4.png)

Reply:
> Aside: I think the ID problem is a great example of how the output of using .closest('form') could've helped identify the right spot to fix in a hypothetical audit :)

But it isn't part of the form. Audit aside, a Form artifact should structure forms accurately, not guessing author intent. Following that logic further ... the previous code snippet should actually place the ill-formed label element in the `formless` category, but give it a property like `.closestParentForm`.

Reply:
Yeah I think that's great it just sounded like your latest comment was cooling on using the `if` in that screenshot and wanted to remove it, but rereading I see what you were saying now. Ignore my latest, carry on :)

Reply:
> the previous code snippet should actually place the ill-formed label element in the formless category

A-ha! So I *was* right the first time to think you didn't want to use `.closest` for the form determination :)

if it's this big a deal, let's drop it entirely for this iteration and we can potentially "upgrade" it to use the parent form in the future?